Allow deletion of pending pods when using PDBS

michaelgugino opened the pull request:
Currently, if you have a PDB set, it is possible for
a pod stuck in pending state to be prevented from
deletion even though there are in fact enough healthy
replicas.

This commit allows pods in Pending state to be removed.

This commit also adds associated unit tests.

Fixes: #80389

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change

/kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
Currently, if you have a PDB set, it is possible for
a pod stuck in pending state to be prevented from
deletion even though there are in fact enough healthy
replicas.
**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->

Fixes: https://github.com/kubernetes/kubernetes/issues/80389

**Special notes for your reviewer**:
Here's where 'healthy' pods are counted: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/disruption/disruption.go#L675

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
When evicting, Pods in Pending state are removed without checking PDBs.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.
Please use the following format for linking documentation or pass the
section below:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

michaelgugino committed:
Allow deletion of pending pods when using PDBS

Currently, if you have a PDB set, it is possible for
a pod stuck in pending state to be prevented from
deletion even though there are in fact enough healthy
replicas.

This commit allows pods in Pending state to be removed.

This commit also adds associated unit tests.

related-bug: #80389

michaelgugino commented:
/sig-apps

michaelgugino commented:
/sig scheduling

michaelgugino commented:
Replaces: https://github.com/kubernetes/kubernetes/pull/81175

michaelgugino commented:
@mortent PTAL

mortent commented:
/ok-to-test
/sig apps
/lgtm

michaelgugino commented:
/test pull-kubernetes-e2e-gce-100-performance

michaelgugino commented:
/test pull-kubernetes-kubemark-e2e-gce-big

mortent commented:
/assign @liggitt

liggitt commented on pkg/registry/core/pod/storage/eviction.go:
```diff
@@ -203,6 +203,15 @@ func (r *EvictionREST) Create(ctx context.Context, name string, obj runtime.Obje
+// canIgnorePDB returns true for pod conditions that allow the pod to be deleted
+// without checking PDBs.
+func canIgnorePDB(pod *api.Pod) bool {
+	if pod.Status.Phase == api.PodSucceeded || pod.Status.Phase == api.PodFailed || pod.Status.Phase == api.PodPending {
```
@kubernetes/sig-node-pr-reviews is `PodPending` the right check here, or do we need to drill down into more detailed aspects (e.g. a `Ready` condition of `True` - "the Pod is able to serve requests and should be added to the load balancing pools of all matching Services;")

michaelgugino replied:
@liggitt I have another patch that is a bit more defensive:

https://github.com/kubernetes/kubernetes/pull/81175

> otherwise we could race evicting with the kubelet marking the pod as ready

Without some kind of distributed locking mechanism, I think we're already exposed to a race here.  We discussed this a bit more in my other PR.

Synopsis of other PR:  Only removing pending pods if we have enough healthy nodes (PDB's consider disrupted nodes, rather than healthy nodes, so we have to perform that calculation ourselves).  This should provide enough buffer against multiple transactions happening at once.

liggitt replied:
the pod object we get on line 119 has a resourceVersion we can use as a precondition in the delete call on line 134 that would ensure we only delete the pod if no other changes were made to it in the meantime. that wasn't important before this PR because succeeded and failed are terminal states.

liggitt replied:
if a change *was* made to the pod in the meantime, the delete call would get a conflict error, and this code could refetch the pod and retry the logic. The RetryOnConflict helper method is used for this in many places.

michaelgugino replied:
@liggitt if we get a conflict error, perhaps we should leave retry up to the client?  Send back the 529 (or whatever code it is), and then the PDB logic will run anew.

How do we ensure the delete call gets this conflict?  I thought that was already built in already?

liggitt replied:
> How do we ensure the delete call gets this conflict? I thought that was already built in already?

set the resourceVersion precondition in deletionOptions if we are making a decision based on the information in a particular revision of the pod (as this PR is introducing)

liggitt commented:
should this also address https://github.com/kubernetes/kubernetes/issues/72320? the symptom (cannot evict pod that is not actually contributing to the "ready pods" count) sounds very similar

michaelgugino commented:
> does/should this also address #72320? the symptom (cannot evict pod that is not actually contributing to the "ready pods" count) sounds very similar

@liggitt I only accounted for the 'pending' pod phase.  CrashLoopBackoff is not a phase.  For that case, I definitely think we would want to ensure there are enough healthy replicas before removing.  In particular, the pod may have already been 'Ready' and previously counted in the PDB calculations, and there may be a delete operation going on elsewhere of another pod when this one fails.  The CrashLoopBackoff pod might still have good data, and the condition might be temporary (eg, dns failure of some needed resource, application exits 1), and it might be bad to remove it without ensure there are enough healthy replicas.

That would probably be best solved with my other patch (modified for crashloopbackoff instead of pending) in conjunction with this patch.

mortent commented:
It seems to me that to avoid situations like the one described in #72320 where the number of pods are higher than the desired number of pods in the PDB and eviction is still blocked for all pods, eviction will need to use the same rules that the disruption controller for deciding whether a pod should be counted against the disruption budget. Currently it is possible that a pod is not counted as available in the disruption controller, but it still decrements the allowed disruptions during evictions.

If we consider that all pods can be evicted regardless of PDB if they don't have the Ready condition set to true (i.e it matches the rules for computing healthy pods in the disruption controller), are there any situations where this will not work? Deletion will be conditioned on the resource version obviously. I don't think there should be any issues with this since a pod without the Ready condition is considered as unavailable in this context, so evicting one will never reduce the number of healthy pods. If the pod would eventually get back to the ready state it might have some impact, but a controller will probably spin up a new replacement pod after eviction anyway.

michaelgugino commented:
> It seems to me that to avoid situations like the one described in #72320 where the number of pods are higher than the desired number of pods in the PDB and eviction is still blocked for all pods, eviction will need to use the same rules that the disruption controller for deciding whether a pod should be counted against the disruption budget. Currently it is possible that a pod is not counted as available in the disruption controller, but it still decrements the allowed disruptions during evictions.
> 
> If we consider that all pods can be evicted regardless of PDB if they don't have the Ready condition set to true (i.e it matches the rules for computing healthy pods in the disruption controller), are there any situations where this will not work? Deletion will be conditioned on the resource version obviously. I don't think there should be any issues with this since a pod without the Ready condition is considered as unavailable in this context, so evicting one will never reduce the number of healthy pods. If the pod would eventually get back to the ready state it might have some impact, but a controller will probably spin up a new replacement pod after eviction anyway.

@mortent 

My concern with deleting anything that's not counted as ready might be data loss if we don't ensure there are in fact enough healthy replicas.  My other patch addresses that case better:
https://github.com/kubernetes/kubernetes/pull/81175

The same method to count healthy replicas is used there (We should probably add a public method somewhere in the PDB code and just import that so it's used consistently).  I think that patch would be good in addition to this patch.

michaelgugino commented on pkg/registry/core/pod/storage/eviction.go:
```diff
@@ -128,9 +128,17 @@ func (r *EvictionREST) Create(ctx context.Context, name string, obj runtime.Obje
+		deletionOptions.Preconditions = &metav1.Preconditions{}
+	}
+	podRV := pod.ResourceVersion
+	deletionOptions.Preconditions.ResourceVersion = &podRV
```
This should work, but it's causing the tests to fail locally for me.  I've spent some time debugging, etcd is incrementing the pod resource version when it gets it retrieves it to verify RV before the delete.  I'm assuming this is only impacting the local testing etcd, but I'll have more debug info coming in an issue soon.

michaelgugino commented:
This is currently blocked by: https://github.com/kubernetes/kubernetes/issues/85485

mortent commented:
So I think the question here is whether there are situations when removing a pod that is not Ready will negatively impact the actual number of available pods. I could argue that if a pod is not Ready, it has already been disrupted (or not finished starting) and therefore removing it will never reduce the number of actually available pods. If so, the status of the PDB is irrelevant. 

An argument against terminating pods that are in the Running phase but not Ready, is that the pod might go back to Ready quickly after a container restart. If it is terminated, it will be up to the pods controller to replace it if necessary and it might take more time.

liggitt commented on pkg/registry/core/pod/storage/eviction.go:
```diff
@@ -130,7 +130,15 @@ func (r *EvictionREST) Create(ctx context.Context, name string, obj runtime.Obje
+			if deletionOptions.Preconditions == nil {
+				deletionOptions.Preconditions = &metav1.Preconditions{}
+			}
+			deletionOptions.Preconditions.ResourceVersion = &pod.ResourceVersion
```
by adding this precondition, you are indicating you want to get a conflict error if anything else has modified the pod since it was fetched on line 119. you would need to handle that conflict error here

michaelgugino replied:
@liggitt what is your preferred way of handling this?  We could return 429 so clients will retry.

michaelgugino replied:
@liggitt Do you want to return the 429 or leave as-is here?  If we do 429, drain client (and probably other clients) will retry here, which is probably what we want.

mortent replied:
I think we can just retry immediately, similar to what we do here https://github.com/kubernetes/kubernetes/blob/b94dca92a83447ad0384471c8a9ce017420714c1/pkg/registry/core/pod/storage/eviction.go#L164
We just need to make sure that we verify all the conditions again after getting the latest version of the pod from the api server.

michaelgugino replied:
It's not entirely clear what that code is doing, the comments look stale.  I'll dig in and see what it's actually doing, hopefully there are tests.

michaelgugino replied:
Okay, I sorted out what that code is doing, it does indeed return a 429 at appropriate times, but it doesn't have good test coverage.

I went ahead and did something similar for the ignorePDB delete call.

michaelgugino commented:
> So I think the question here is whether there are situations when removing a pod that is not Ready will negatively impact the actual number of available pods. I could argue that if a pod is not Ready, it has already been disrupted (or not finished starting) and therefore removing it will never reduce the number of actually available pods. If so, the status of the PDB is irrelevant.

@mortent 

I think if we always ensure we have the minimum healthy replicas, we can do no wrong.  We could possibly add other pod phases or states here, but my preference would be to check we actually have enough replicas.  If we don't have enough replicas, that's probably something that requires intervention.


> An argument against terminating pods that are in the Running phase but not Ready, is that the pod might go back to Ready quickly after a container restart. If it is terminated, it will be up to the pods controller to replace it if necessary and it might take more time.

In the context of eviction, this is probably okay.  Someone has requested for that pod to go away for one reason or another (eg, drain).  It's certainly expected that the pod will have to be started elsewhere.  As long as we don't violate PDBs, it shouldn't matter what state the particular pod is in.

If we're talking about a pod that could potentially flap in and out of readiness, we should probably check PDB health before proceeding.  AFAIK, when a pod is pending, it was never 'ready' so it's not a candidate for flapping from ready.  I'm unsure about other phases.  I want to be as careful as possible when it comes to other phases so we can honor PDBs as much as possible.

mortent commented:
> > So I think the question here is whether there are situations when removing a pod that is not Ready will negatively impact the actual number of available pods. I could argue that if a pod is not Ready, it has already been disrupted (or not finished starting) and therefore removing it will never reduce the number of actually available pods. If so, the status of the PDB is irrelevant.
> 
> @mortent
> 
> I think if we always ensure we have the minimum healthy replicas, we can do no wrong. We could possibly add other pod phases or states here, but my preference would be to check we actually have enough replicas. If we don't have enough replicas, that's probably something that requires intervention.

We do want to keep the minimum number of healthy replicas, but the issue here is what do we do with unhealthy replicas. Are there any situations where terminating an unhealthy pod would result in things being worse off than if the unhealthy pod were left running? The current situation where we don't terminate unhealthy replicas because of PDBs leads to situations where there are more replicas than needed by the PDB, but none of them can be terminated. This comment seems to be a good example https://github.com/kubernetes/kubernetes/issues/72320#issuecomment-557961164

> 
> > An argument against terminating pods that are in the Running phase but not Ready, is that the pod might go back to Ready quickly after a container restart. If it is terminated, it will be up to the pods controller to replace it if necessary and it might take more time.
> 
> In the context of eviction, this is probably okay. Someone has requested for that pod to go away for one reason or another (eg, drain). It's certainly expected that the pod will have to be started elsewhere. As long as we don't violate PDBs, it shouldn't matter what state the particular pod is in.

Good point. I agree

> 
> If we're talking about a pod that could potentially flap in and out of readiness, we should probably check PDB health before proceeding. AFAIK, when a pod is pending, it was never 'ready' so it's not a candidate for flapping from ready. I'm unsure about other phases. I want to be as careful as possible when it comes to other phases so we can honor PDBs as much as possible.

I agree that for pending pods it should be fine, as long as we use the conditional delete where we make sure the pod has not transitioned to the Running phase by the time it is deleted. I just think pods that are not Ready, for whatever reason, should also be safe to delete as long as we make sure they have not transitioned to Ready by the time they are deleted.

This might be a good discussion for the next sig-apps meeting to get input from a broader group. I'll see if we can get it on the agenda.

michaelgugino commented:
@mortent 


> We do want to keep the minimum number of healthy replicas, but the issue here is what do we do with unhealthy replicas. Are there any situations where terminating an unhealthy pod would result in things being worse off than if the unhealthy pod were left running? The current situation where we don't terminate unhealthy replicas because of PDBs leads to situations where there are more replicas than needed by the PDB, but none of them can be terminated. This comment seems to be a good example [#72320 (comment)](https://github.com/kubernetes/kubernetes/issues/72320#issuecomment-557961164)

So, I think this is better addressed by something other than PDBs.  You can't wish
to respect PDBs, and also wish to not respect PDBs, those things are mutually exclusive.

Firstly, I think this is a well-defined problem at this point.  Build monitoring and
alerting around prolonged PDB failures.

Something we could do: maybe we can indicate on PDB's the last
transition time to 'unhealthy' status, and if it's above X configurable time, ignore it.
This would be done on the drain client-side of the library, rather than the eviction server-side
IMO.

Here's another alternative:  After X time trying to drain a node, and you're certain
you *really* want to ignore pdbs, run drain with this patch: https://github.com/kubernetes/kubernetes/pull/85571


> I agree that for pending pods it should be fine, as long as we use the conditional delete where we make sure the pod has not transitioned to the Running phase by the time it is deleted. I just think pods that are not Ready, for whatever reason, should also be safe to delete as long as we make sure they have not transitioned to Ready by the time they are deleted.

I'm concerned about temporary 'un-readiness' that might be resolved.  The application can be alive and doing things and not be 'ready' as readiness probes are configurable.  Say you have two databases rebuilding indexes and they go 'unready' during this process so they're not serving requests, but they're otherwise okay.  Deleting two of them would probably be bad.

> This might be a good discussion for the next sig-apps meeting to get input from a broader group. I'll see if we can get it on the agenda.

Next one at Dec 2?  I should be able to make that.

mortent commented:
/reopen
I think this got closed by mistake when #85516 merged.

mortent commented:
> @mortent
> 
> > We do want to keep the minimum number of healthy replicas, but the issue here is what do we do with unhealthy replicas. Are there any situations where terminating an unhealthy pod would result in things being worse off than if the unhealthy pod were left running? The current situation where we don't terminate unhealthy replicas because of PDBs leads to situations where there are more replicas than needed by the PDB, but none of them can be terminated. This comment seems to be a good example [#72320 (comment)](https://github.com/kubernetes/kubernetes/issues/72320#issuecomment-557961164)
> 
> So, I think this is better addressed by something other than PDBs. You can't wish
> to respect PDBs, and also wish to not respect PDBs, those things are mutually exclusive.
> 
> Firstly, I think this is a well-defined problem at this point. Build monitoring and
> alerting around prolonged PDB failures.
> 
> Something we could do: maybe we can indicate on PDB's the last
> transition time to 'unhealthy' status, and if it's above X configurable time, ignore it.
> This would be done on the drain client-side of the library, rather than the eviction server-side
> IMO.
> 
> Here's another alternative: After X time trying to drain a node, and you're certain
> you _really_ want to ignore pdbs, run drain with this patch: #85571
> 
> > I agree that for pending pods it should be fine, as long as we use the conditional delete where we make sure the pod has not transitioned to the Running phase by the time it is deleted. I just think pods that are not Ready, for whatever reason, should also be safe to delete as long as we make sure they have not transitioned to Ready by the time they are deleted.
> 
> I'm concerned about temporary 'un-readiness' that might be resolved. The application can be alive and doing things and not be 'ready' as readiness probes are configurable. Say you have two databases rebuilding indexes and they go 'unready' during this process so they're not serving requests, but they're otherwise okay. Deleting two of them would probably be bad.
> 
> > This might be a good discussion for the next sig-apps meeting to get input from a broader group. I'll see if we can get it on the agenda.
> 
> Next one at Dec 2? I should be able to make that.

So I don't think this a question about respecting PDBs or not. It is a question of when a pod should already be considered disrupted and therefore can be terminated. This seems pretty straight-forward for pending pods. Terminating non-ready pods is a more complex situation, but I think it is worth discussing. It is on the agenda for next sig-apps on Dec 2.

mortent reviewed (approved):
/lgtm

michaelgugino commented on pkg/registry/core/pod/storage/eviction.go:
```diff
@@ -46,24 +45,25 @@ const (
 }
 
 // EvictionREST implements the REST endpoint for evicting pods from nodes
 type EvictionREST struct {
```
I grepped through the codebase, I don't see this being used in any other packages and I don't know why it  would be.  This change is needed to support mocking storage.

michaelgugino replied:
I didn't see EvictionREST being used outside of this package, and the newEvictionStorage constructor is private.  As you can see in the constructor method, there is not any change required when instantiating EvictionStorage, so it should be entirely transparent if someone is importing this package elsewhere.

michaelgugino commented:
@mortent @liggitt 

Please take a look.  I added some tests for the critical section in the delete-retry logic.  I think it's in pretty good shape now, and the mock I created could be used to help test the retry section of the existing code in a follow up.

mortent reviewed (commented):
Looks good to me.

liggitt commented:
the various PDB PRs are on my list to work through once 1.17.0 work wraps up, later this week or first thing next week.

michaelgugino commented:
@liggitt any updates?

mortent commented:
@liggitt Can you take a look at this one when you have a chance.

michaelgugino commented:
@liggitt ping.  Any chance you'll get to this one soon?

liggitt commented on pkg/registry/core/pod/storage/eviction.go:
```diff
@@ -130,13 +130,46 @@ func (r *EvictionREST) Create(ctx context.Context, name string, obj runtime.Obje
+		err := retry.RetryOnConflict(EvictionsRetry, func() error {
+			if deleteRefresh {
+				// If we hit a conflict error, get the latest pod
+				obj, err = r.store.Get(ctx, eviction.Name, &metav1.GetOptions{})
```
You're getting the latest pod here, but not updating the deletionOptions.Preconditions.ResourceVersion precondition, which means you will continue getting the same conflict error. Need a test case to cover this scenario.

liggitt replied:
we actually need two additional tests of the conflict path here:
1. canIgnorePDB=true, shouldEnforceResourceVersion=true should update the precondition resource version
2. canIgnorePDB=true, shouldEnforceResourceVersion=false should use the original resource version precondition from the eviction request
3. canIgnorePDB=false should continue to PDBs and use the original resource version precondition from the eviction request

liggitt commented on pkg/registry/core/pod/storage/eviction.go:
```diff
@@ -130,13 +130,46 @@ func (r *EvictionREST) Create(ctx context.Context, name string, obj runtime.Obje
+			if deletionOptions.Preconditions == nil {
+				deletionOptions.Preconditions = &metav1.Preconditions{}
+			}
+			deletionOptions.Preconditions.ResourceVersion = &pod.ResourceVersion
```
You're modifying `deletionOptions.Preconditions.ResourceVersion` here in a way that is correct for the canIgnorePDB case, but is not correct if we end up in the continueToPDBs case. In that scenario, we want the original deletionOptions to be used. Need a test case to cover this scenario:
1. eviction request is received with no resourceVersion precondition
2. pod at rv=1 is pending, so canIgnorePDBs is true
3. deletionOptions gets a rv=1 precondition added
4. initial eviction attempt gets a conflict error
5. next pod get returns non-pending pod at rv=2, so canIgnorePDBs is false, so we continue to PDBs
6. PDB allows eviction
7. Pod is deleted successfully

Currently, that scenario would fail in step 7 with a conflict error because of the leftover rv=1 precondition

michaelgugino commented:
/retest

michaelgugino commented:
/retest

michaelgugino commented:
/test pull-kubernetes-e2e-gce

michaelgugino commented:
/retest

michaelgugino commented:
/test pull-kubernetes-e2e-gce

michaelgugino commented:
/test pull-kubernetes-e2e-gce

michaelgugino commented:
@liggitt PTAL.  I have squashed the commits, added the tests/features you requested.  The tests are little hacky, but they work and aren't too bad if you take the time to digest what they're doing.

deads2k commented on pkg/registry/core/pod/storage/eviction.go:
```diff
@@ -130,13 +130,55 @@ func (r *EvictionREST) Create(ctx context.Context, name string, obj runtime.Obje
+		deleteRefresh := false
+		continueToPDBs := false
+		// Preserve current deletionOptions if we need to fall through to checking PDBs later.
+		preservedDeletionOptions := deletionOptions.DeepCopy()
```
minor: rename `deleteOptions` to `originalDeleteOptions` and do something like

```go
retry.RetryOnConflict{
  // avoid mutating original delete options.
  currDeleteOptions := originalDeleteOptions.DeepCopy()
}
```

The reassignment of bits down below is making me cross-eyed.

deads2k commented on pkg/registry/core/pod/storage/eviction.go:
```diff
@@ -130,13 +130,55 @@ func (r *EvictionREST) Create(ctx context.Context, name string, obj runtime.Obje
+		continueToPDBs := false
+		// Preserve current deletionOptions if we need to fall through to checking PDBs later.
+		preservedDeletionOptions := deletionOptions.DeepCopy()
+		err := retry.RetryOnConflict(EvictionsRetry, func() error {
```
it's not completely obvious to me why we need retries on conflicts.  Is this conflicting often enough that it warrants the complexity?  I worry that the intent of this PR was simply to add an ` || pod.Status.phase == pending` and it's gotten stuck in review over complicated retry logic we didn't need.

liggitt replied:
Previously, the decision about whether a pod could be evicted was based on a terminal state, so no subsequent update to the pod could invalidate that decision.

This PR adds the possibility of basing that decision on something that can be changed by an update, and introduces the possibility of a race between eviction deleting the pod and the node starting the pod. Example flow:
1. pod is pending
2. eviction request is submitted
2. eviction handler decides to evict based on the pod pending
3. pod is scheduled
4. node observes the scheduled pod and starts the process
5. eviction handler deletes the pod

That means we need to add a precondition to ensure the eviction is performed on the same data we based our decision on. That introduces the possibility of conflict errors, which we should handle internally since clients submitting unconditional evictions aren't expecting to receive conflicts.

deads2k replied:
> That introduces the possibility of conflict errors, which we should handle internally since clients submitting unconditional evictions aren't expecting to receive conflicts.

This bit needs to be a comment in the code.  Also, it seems that conflicts from deletes are reasonable given finalizers, but if conflict retry is important to you, so be it.

deads2k commented on pkg/registry/core/pod/storage/eviction.go:
```diff
@@ -130,13 +130,55 @@ func (r *EvictionREST) Create(ctx context.Context, name string, obj runtime.Obje
-		_, _, err = r.store.Delete(ctx, eviction.Name, rest.ValidateAllObjectFunc, deletionOptions)
-		if err != nil {
-			return nil, err
+	if canIgnorePDB(pod) {
```
since you're doing a simple GET down below and rechecking this condition it seems like you could simply enter the retryOnlyConflict and extract a `*core.Pod` out of it for later use.

deads2k commented on pkg/registry/core/pod/storage/eviction.go:
```diff
@@ -130,13 +130,55 @@ func (r *EvictionREST) Create(ctx context.Context, name string, obj runtime.Obje
+		// Preserve current deletionOptions if we need to fall through to checking PDBs later.
+		preservedDeletionOptions := deletionOptions.DeepCopy()
+		err := retry.RetryOnConflict(EvictionsRetry, func() error {
+			if deleteRefresh {
```
that would make this unconditional

deads2k commented on pkg/registry/core/pod/storage/eviction.go:
```diff
@@ -130,13 +130,55 @@ func (r *EvictionREST) Create(ctx context.Context, name string, obj runtime.Obje
+			}
+			return nil
+		})
+		if !continueToPDBs {
```
this is actually an `if successfullyDeletedPod` condition check, right?

deads2k committed:
reduce complexity in pdb refactor

michaelgugino commented:
/retest

michaelgugino commented:
/retest

michaelgugino commented:
/retest

michaelgugino commented:
Looks like pull-kubernetes-e2e-kind is flaking more often than not.  Opened https://github.com/kubernetes/kubernetes/issues/91128

michaelgugino commented:
/test pull-kubernetes-e2e-kind

michaelgugino commented:
@liggitt @mortent I think this is ready for some labels if you have the time.

liggitt commented on pkg/registry/core/pod/storage/eviction.go:
```diff
@@ -203,6 +235,27 @@ func (r *EvictionREST) Create(ctx context.Context, name string, obj runtime.Obje
+
+func shouldEnforceResourceVersion(pod *api.Pod) bool {
+	// Only pods that may be included as health in PDBs in the future need to be checked.
+	if pod.Status.Phase == api.PodPending {
```
is this only run on pods that pass `canIgnorePDB`? would it be clearer to default to enforcing resourceVersion except in terminal cases, e.g.:

```go
if pod.Status.Phase == api.PodSucceeded || pod.Status.Phase == api.PodFailed {
  return false
}
return true
```

michaelgugino replied:
Yes, this is only run on pods that pass canIgnorePDB.  I think what you're saying makes sense.  You need to specifically opt-out of enforcing ResourceVersion at this point, and default to enforcing otherwise.

I think that's pretty minor, I could do a follow up if you don't think it's a blocker.

deads2k replied:
> Yes, this is only run on pods that pass canIgnorePDB. I think what you're saying makes sense. You need to specifically opt-out of enforcing ResourceVersion at this point, and default to enforcing otherwise.
> 
> I think that's pretty minor, I could do a follow up if you don't think it's a blocker.

it's an easy, minor comment.  Let's just get it cleaned up.

liggitt commented on pkg/registry/core/pod/storage/eviction.go:
```diff
@@ -111,33 +111,65 @@ func (r *EvictionREST) Create(ctx context.Context, name string, obj runtime.Obje
 		_, _, err = r.store.Delete(ctx, eviction.Name, rest.ValidateAllObjectFunc, deletionOptions)
 		if err != nil {
-			return nil, err
+			return err
```
if we get a conflict error and originalDeleteOptions had a resourceVersion specified, we should bail... retrying won't do anything good for us.

liggitt commented on pkg/registry/core/pod/storage/eviction.go:
```diff
@@ -111,33 +111,65 @@ func (r *EvictionREST) Create(ctx context.Context, name string, obj runtime.Obje
+	switch {
+	case err != nil:
+		// this can happen in cases where the PDB can be ignored, but there was a problem issuing the pod delete:
+		// maybe we conflicted two many times or we didn't have permission or something else weird.
```
```suggestion
		// maybe we conflicted too many times or we didn't have permission or something else weird.
```

liggitt commented:
a couple nits and one substantive comment on retry behavior when the original delete options specify a resource version

michaelgugino committed:
Quit retrying early with user supplied resourceVersion

This commit also updates tests and comments.

michaelgugino commented on pkg/registry/core/pod/storage/eviction.go:
```diff
@@ -104,40 +104,85 @@ func propagateDryRun(eviction *policy.Eviction, options *metav1.CreateOptions) (
 		_, _, err = r.store.Delete(ctx, eviction.Name, rest.ValidateAllObjectFunc, deletionOptions)
 		if err != nil {
-			return nil, err
+			if kerrors.IsConflict(err) && !resourceVersionIsUnset(originalDeleteOptions) {
```
If you look at the 3rd commit, you can see the test case that covers this.  We are detecting we only get 1 attempt.

michaelgugino commented on pkg/registry/core/pod/storage/eviction.go:
```diff
@@ -104,40 +104,85 @@ func propagateDryRun(eviction *policy.Eviction, options *metav1.CreateOptions) (
+				// We hit a conflict here for a requestor-specified resourceVersion.
+				// We save the error and return something else.
+				unrecoverableConflict = &err
+				return errors.New("Unrecoverable Conflict")
```
Need to create a new error type here otherwise there's no way to break the retry loop that I'm aware of.  kerrors.IsConflict also works with wrapped errors according to docs, so that possibly won't work either.

liggitt replied:
see suggestion at https://github.com/kubernetes/kubernetes/pull/83906#discussion_r432075989

michaelgugino commented:
@liggitt PTAL, I think if you don't have any objections to how I handled the errors, I think it's ready.

liggitt commented on pkg/registry/core/pod/storage/eviction.go:
```diff
@@ -104,40 +104,85 @@ func propagateDryRun(eviction *policy.Eviction, options *metav1.CreateOptions) (
-	if pod.Status.Phase == api.PodSucceeded || pod.Status.Phase == api.PodFailed {
+	var pod *api.Pod
+	deletedPod := false
+	var unrecoverableConflict *error
+	err = retry.RetryOnConflict(EvictionsRetry, func() error {
```
would this be simpler:

```go
// by default, retry conflict errors
shouldRetry := kerrors.IsConflict
if !resourceVersionIsUnset(originalDeleteOptions) {
	// if the original options included a resourceVersion precondition, don't retry
	shouldRetry = func(err error) bool {return false}
}

err = retry.OnError(EvictionsRetry, shouldRetry, func() error {
```

michaelgugino replied:
That retry library is handy.  Great suggestion.

liggitt commented:
/lgtm
/approve

michaelgugino commented:
/test pull-kubernetes-e2e-kind

michaelgugino commented:
/test pull-kubernetes-e2e-kind

michaelgugino commented:
/test pull-kubernetes-e2e-kind

michaelgugino commented:
/test pull-kubernetes-conformance-kind-ga-only-parallel

michaelgugino commented:
/test pull-kubernetes-e2e-kind

michaelgugino commented:
/test pull-kubernetes-e2e-kind

michaelgugino commented:
/test pull-kubernetes-e2e-kind

michaelgugino commented:
/test pull-kubernetes-e2e-kind
